node_modules/oae-core/documentpreview/js/documentpreview.js

```diff
@@ -71,6 +71,95 @@ define(['jquery', 'underscore', 'oae.core', 'lazyload'], function($, _, oae) {
+            // scrolling correctly.
+            page.height = page.$el.height();
+            page.marginTop = CONTENT_PAGE_SPACING;
+        };
+
+        /**
+         * Render the HTML content for a page
+         *
+         * @param  {Object}         page                Page object representing the page that will be loaded
+         * @param  {String}         html                HTML content for the page
+         */
+        var renderPageHtml = function(page, html) {
+            // Add the HTML content (overwriting the loading indicator)
+            page.$html = $('.documentpreview-content-html-page', page.$el).html(html);
+
+            // Update the height now that we have real content
+            page.height = page.$el.height();
+
+            // If the page being loaded is the first page of the document, we adjust
+            // the zoom level based on the available space to make sure that the page
+            // fully fits inside of the viewer
+            if (page.pageNumber === 1) {
+                determineZoomLevel(page);
+            }
+
+            // Adjust the page display and any pages following it to reflect the
+            // current zoom level
+            zoomPages(page.pageNumber);
+        };
+
+        /**
+         * Render the text content for a page (for screen readers)
+         *
+         * @param  {Object}         page                Page object representing the page that will be loaded
+         * @param  {String}         text                Text content for the page
+         */
+        var renderPageText = function(page, text) {
+            // Add the text content
+            page.$text = $('.documentpreview-content-text-page', page.$el)
+            $('textarea', page.$text).text(text);
+        };
+
+        /**
+         * Handle page loading errors to recover from signature expiry
+         *
+         * @param  {Object}         page                Page object representing the page that will be loaded
+         * @param  {Function}       callback            Standard callback function
+         * @param  {Boolean}        _isRetryAttempt    `true` if we attempted to load this page previously
+         * @param  {Object}         jqXHR               jQuery XHR from the failed load attempt
+         */
+        var handlePageLoadError = function(page, callback, _isRetryAttempt, jqXHR) {
+            // If the user has been reading a document for a while, it's possible that the signature expired
+            // If we could not load the page due to a 401 (access denied) and it's the first time we tried loading this
+            // particular page, we refresh the signature and try again
+            if (jqXHR.status === 401 && !_isRetryAttempt) {
+                refreshSignature(function(err) {
+                    if (err) {
+                        callback(err);
+                    }
+
+                    // Try loading the page again
+                    loadPage(page, callback, true);
```

Comment:
I think there's currently a chance that we'd load the same page twice as the signature can be refreshed twice. The code would detect that the signature is out of date, and depending on the timing of the requests, could call handlePageLoadError twice.